listenbrainz_spark/recommendations/train_models.py

```diff
@@ -200,9 +198,8 @@ def main():
     logging.info('Saving model...')
     t0 = time()
     try:
-        path = os.path.join('/', 'data', 'listenbrainz', 'recommendation-engine', 'best-model', '{}' \
-        .format(best_model_metadata['model_id']))
-        model.model.save(listenbrainz_spark.context, config.HDFS_CLUSTER_URI + path)
+        _path = path.DATA_DIR + '/' + best_model_metadata['model_id']
+        model.model.save(listenbrainz_spark.context, config.HDFS_CLUSTER_URI + _path)
```

Comment:
os.path.join in the line here and the one above?

Reply:
os.path.join doesn't work with HDFS paths.

Reply:
The problem here is not an HDFS path mainly. For instance, I have a dir path as:
` DATA_DIR = os.path.join('/', 'listenbrainz', 'data') `
DATA_DIR = '/data/listenbrainz'
Now, when I want to access a file from DATA_DIR I would have to specify the full path which is:
 'hdfs://hadoop-master:9000/data/listenbrainz'

HDFS_CLUSTER_URI =  'hdfs://hadoop-master:9000'
`FINAL_PATH = os.path.join(HDFS_CLUSTER_URI, DATA_DIR)` 
FINAL_PATH  = '/data/listenbrainz' which is defintely not what we want since os.path.join ignores everything before a '/'
That is why we have used '+'. Note that FINAL_PATH is an url and not exactly a path.

We can do something like this if os.path.join must be used
`os.path.join('hdfs://hadoop-master:9000', 'data', 'listenbrainz')`
but there are places where we just want DATA_DIR and not the full URL. In order to maintain flexibility I have kept DIRs and URLs diff.